src/crimson/net/Errors.cc

```diff
@@ -61,7 +61,11 @@ const std::error_category& net_category()
     }
 
     bool equivalent(int code, const std::error_condition& cond) const noexcept override {
-      switch (static_cast<error>(code)) {
+      const auto &err = static_cast<error>(code);
+      if (cond == err) {
```

Comment:
can you clarify why this is needed? what comparison was failing without this part?

Reply:
The direct reason is that (e.g.) `e.code() == error::read_eof` will result in `false` even if they appear same from debugging.

Under the hood (IMO), the rh `error::read_eof` will be implicitly converted to type `std::error_condition` because of `struct is_error_condition_enum<ceph::net::error> : public true_type {};` To compare lh `std::error_code` with rh `std::error_condition`, the legacy `equivalent()` will return `false` even if the `code` matches `cond`.

Fix: return `true` if `cond` equals `static_cast<error>(code)` in the new `equivalent()`, note the rh `static_cast<error>(code)` will also be implicitly converted to `std::error_condition`. So no need to worry about infinite recursion.

Reply:
Then the multiple `close()` issue pops out, previously it won't be called in `read_tags_until_next_message()` in any condition.

Reply:
thanks! sorry for the delays, i wanted to do some extra testing here. it looks like the default implementations of `std::error_category::equivalent()` both do the right thing here, so i'd rather call them for this part

i opened https://github.com/ceph/ceph/pull/23844 as an example. if that works for you, you're welcome to pull those commits into this pr, or we can merge that one separately - up to you

Reply:
thanks for the fix! I'll pull #23844 into this pr, because the multiple-call of `SocketConnection::close()` depends on this fix.

Reply:
done.